test/e2e/dra/README.md

```diff
@@ -36,6 +36,180 @@ $ kind build node-image --image dra/node:latest $(pwd)
+
+You can now create your first workload cluster by running the following:
+
+  clusterctl generate cluster [name] --kubernetes-version [version] | kubectl apply -f -
+
+# Patching in the configuration for pod security conflicts with
+# extraArgs for apisever due to this JSON patch:
+#     - jsonPatches:
+#      - op: add
+#        path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs
+#        value:
+#         admission-control-config-file: /etc/kubernetes/kube-apiserver-admission-pss.yaml
+
+$ POD_SECURITY_STANDARD_ENABLED=false clusterctl generate cluster capi-quickstart --flavor development   --kubernetes-version $version --control-plane-machine-count=1   --worker-machine-count=1 >cluster-latest.yaml
+
+# MachinePools are experimental, so this patch removes them from the cluster.
+# Autoscaling will be done through MachineDeployments.
+
+$ cat cluster.patch <<EOF
+*** cluster-1.30.0.yaml	2024-01-15 18:03:14.507591268 +0100
+--- cluster-1.30.0-dra-II.yaml	2024-01-15 19:28:02.918296629 +0100
+***************
+*** 250,255 ****
+--- 250,258 ----
+        kubeadmConfigSpec:
+          clusterConfiguration:
+            apiServer:
++             extraArgs:
++               runtime-config: resource.k8s.io/v1alpha2=true
++               feature-gates: DynamicResourceAllocation=true,ContextualLogging=true
+              certSANs:
+              - localhost
+              - 127.0.0.1
+***************
+*** 258,263 ****
+--- 261,271 ----
+            controllerManager:
+              extraArgs:
+                enable-hostpath-provisioner: "true"
++               feature-gates: DynamicResourceAllocation=true,ContextualLogging=true
++           scheduler:
++             extraArgs:
++               feature-gates: DynamicResourceAllocation=true,ContextualLogging=true
++         # TODO: enable features in kubelet
+          initConfiguration:
+            nodeRegistration: {}
+          joinConfiguration:
+***************
+*** 306,312 ****
+    template:
+      spec:
+        joinConfiguration:
+!         nodeRegistration: {}
+  ---
+  apiVersion: cluster.x-k8s.io/v1beta1
+  kind: Cluster
+--- 314,328 ----
+    template:
+      spec:
+        joinConfiguration:
+!         nodeRegistration:
+!           kubeletExtraArgs:
+!             feature-gates: DynamicResourceAllocation=true,ContextualLogging=true
+!       initConfiguration:
+!         nodeRegistration:
+!           kubeletExtraArgs:
+!             # Currently ignored and/or overwritten?
+!             # /var/run/kubeadm/kubeadm.yaml on the control plane container doesn't have it.
+!             feature-gates: DynamicResourceAllocation=true,ContextualLogging=true
+  ---
+  apiVersion: cluster.x-k8s.io/v1beta1
+  kind: Cluster
+***************
+*** 346,352 ****
+        - class: default-worker
+          name: md-0
+          replicas: 1
+-       machinePools:
+-       - class: default-worker
+-         name: mp-0
+-         replicas: 1
```

Comment:
I think this should be dropped (so that the autoscaler knows that it can control the field)
```suggestion
```

Reply:
@marquiz: agreed?

I pushed [an update](https://github.com/kubernetes/kubernetes/compare/ec6e2b218a2bd629088747d04bdf64bca23bb621..7c6c17cca538fa1201a5252ecec79b0d5b31bac5) that addresses the other points.

Sorry for the delay, I had to put this aside for a while to focus on 1.30. It's still useful to have for 1.31, because that is where we hopefully will have cluster autoscaler support.

/cc @towca

Reply:
@pohly sorry my comment was off-by-a-few-lines. I think the `replicas` field should be removed for the `machineDeployments`

Reply:
As it stands, the cluster comes up with one worker node. Then autoscaler can scale up or down. I prefer that over not bringing up any worker node initially because then a problem with that only surfaces later.

Your comment was "so that the autoscaler knows that it can control the field" - I don't think that setting the initial value prevents that.

Reply:
IIRC in my testing it did. When the autoscaler sees that it's set it determines that it's controlled by some other entity and refuses to act on it. It default to 1 (if not set), IIRC

Reply:
You are right, the instructions are not enough to actually make the autoscaler do anything. It finds no node groups. But did you really get it to work as you suggested above?

What is missing is something else, the annotations on the MachineDeployment:
https://cluster-api.sigs.k8s.io/tasks/automated-machine-management/autoscaling#enabling-autoscaling

@sbueringer: any suggestion how to get those annotations added automatically to the MachineDeployment?

Do we need `--node-group-auto-discovery=clusterapi:clusterName=capi-quickstart` or is it enabled by default?

https://cluster-api.sigs.k8s.io/tasks/automated-machine-management/autoscaling#configuring-node-group-auto-discovery says "you must configure node group auto discovery" but https://cluster-api.sigs.k8s.io/tasks/automated-machine-management/autoscaling#enabling-autoscaling says "The autoscaler will monitor any MachineSet, MachineDeployment, or MachinePool containing both of these annotations."

When the cluster comes up after following the instructions in this README.md, it has a generated machine deployment with a variable name. I could come up with a kubectl invocation that adds the annotations, but it would be nicer to include that in the cluster configuration.

Also, does the cloud provider in use here (Docker, from `clusterctl init --infrastructure docker`) support scale from zero automatically? https://cluster-api.sigs.k8s.io/tasks/automated-machine-management/autoscaling#scale-from-zero-support documents some additional annotations, but it is not clear if they are needed.

Reply:
> @sbueringer: any suggestion how to get those annotations added automatically to the MachineDeployment?

You can patch:
```yaml
        - class: default-worker
          name: md-0
          replicas: 1
```
to be this instead:
```yaml
        - class: default-worker
          name: md-0
          metadata:
            annotations:
              cluster.x-k8s.io/cluster-api-autoscaler-node-group-min-size: "1"
              cluster.x-k8s.io/cluster-api-autoscaler-node-group-max-size: "3"
```

The result is:
* the annotations will be set on the MachineDeployment
* the MachineDeployment webhook will pick the value of min-size as initial value for MD.spec.replicas

> Do we need --node-group-auto-discovery=clusterapi:clusterName=capi-quickstart or is it enabled by default?

I'm not sure if you have to set `--node-group-auto-discovery`. In our e2e test we do it, but we also have other tests running in other namespaces. You can give it a try without it.

> Also, does the cloud provider in use here (Docker, from clusterctl init --infrastructure docker) support scale from zero automatically?

CAPD falls under:

> If your Cluster API provider does not have support for scaling from zero, you may still use this feature through the capacity annotations.

So you'll have to set the annotations. Also in CAPD you can't really define via DockerMachine which size your "Machine" has. But you can check on the created Nodes via Nodes.status.capacity what the capacity of a Node is. Not really sure where that is coming from though.

Reply:
Reminder to self: this still needs to be included in the instructions.